### ISSUE TITLE
Using @dataclass(eq=False) changes returned type of __eq__ to boolean

### ISSUE THREAD

**Issue by fabiencelier** [microsoft/pyright#1132]
**Describe the bug**

I have custom implementation of `__eq__` method that does not return bool. When using a dataclass I use the annotation `@dataclass(eq=False)`.
Pyright infers the returned type of the `__eq__` to `bool`

**To Reproduce**

This is a minimal example to reproduce the issue.
Note that if you remove the `@dataclass` annotations Pyright does not detect any issue.

```
from __future__ import annotations
from dataclasses import dataclass


class B:
    pass


@dataclass(eq=False)
class A:
    def __eq__(self, other: A) -> B:
        return B()


def test(a: A) -> B:
    operation = a == a
    return operation
```

**VS Code extension or command-line**
both with pyright version 1.1.82
Note that it did not raise any error in 1.1.62

**Comment by fabiencelier** [comment]
This was introduced in 1.1.78 by this commit https://github.com/microsoft/pyright/commit/89b869c7b338e3ebffaa24a42b55ad15cf2f1b98

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is now addressed in pyright 1.1.87, which I just published. It will also be included in the next release of pylance.

**Comment by fabiencelier** [comment]
Hello, thanks for the fix !
It does solve the issue with `==` and `!=` but I still have the issue with comparaison.

If I take back the previous example:

```python
from __future__ import annotations
from dataclasses import dataclass


class B:
    pass


@dataclass(eq=False)
class A:
    def __eq__(self, other: A) -> B:
        return B()

    def __lt__(self, other: A) -> B:
        return B()


def test_eq(a: A) -> B: # this works
    operation = a == a
    return operation


def test_lt(a: A) -> B: # this does not work
    operation = a < a
    return operation
```
With pyright 1.1.87

<pre>
  25:12 - error: Expression of type "bool" cannot be assigned to return type "B"
    "bool" is incompatible with "B" (reportGeneralTypeIssues)
</pre>

**Comment by fabiencelier** [comment]
Oh it works if I set order to False 

```python
@dataclass(eq=False, order=False)
```
However if I understand the [dataclasses doc](https://docs.python.org/3/library/dataclasses.html#module-level-decorators-classes-and-functions) correctly order is supposed to default to False.

**Comment by erictraut** [comment]
Ah, I missed the fact that `eq` defaults to True and `order` defaults to False.

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is now addressed in Pyright 1.1.88, which I just published. It will also be included in the next release of Pylance.
